components/ws-daemon/pkg/controller/controller.go

```diff
@@ -64,170 +73,219 @@ func NewWorkspaceController(c client.Client, opts WorkspaceControllerOpts) (*Wor
-						Type:               string(workspacev1.WorkspaceConditionContentReady),
-						Status:             metav1.ConditionTrue,
-						Message:            "content is ready",
-						LastTransitionTime: metav1.Now(),
-					})
-				}
-
-				return wsc.Status().Update(ctx, &workspace)
-			})
-			if err != nil {
-				return ctrl.Result{}, err
-			}
-		}
+	if workspace.Status.Phase == workspacev1.WorkspacePhaseStopping {
+		result, err = wsc.handleWorkspaceStop(ctx, &workspace, req)
+		return result, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
-func (wsc *WorkspaceController) creator(ws *workspacev1.Workspace, init *csapi.WorkspaceInitializer) session.WorkspaceFactory {
-	var checkoutLocation string
-	allLocations := csapi.GetCheckoutLocationsFromInitializer(init)
-	if len(allLocations) > 0 {
-		checkoutLocation = allLocations[0]
-	}
-
-	serviceDirName := ws.Name + "-daemon"
-	return func(ctx context.Context, location string) (res *session.Workspace, err error) {
-		return &session.Workspace{
-			Location:              location,
-			CheckoutLocation:      checkoutLocation,
-			CreatedAt:             time.Now(),
-			Owner:                 ws.Spec.Ownership.Owner,
-			WorkspaceID:           ws.Spec.Ownership.WorkspaceID,
-			InstanceID:            ws.Name,
-			FullWorkspaceBackup:   false,
-			PersistentVolumeClaim: false,
-			RemoteStorageDisabled: ws.Status.Headless,
-			// StorageQuota:          int(req.StorageQuotaBytes),
-
-			ServiceLocDaemon: filepath.Join(wsc.opts.ContentConfig.WorkingArea, serviceDirName),
-			ServiceLocNode:   filepath.Join(wsc.opts.ContentConfig.WorkingAreaNode, serviceDirName),
-		}, nil
+func (wsc *WorkspaceController) handleWorkspaceInit(ctx context.Context, ws *workspacev1.Workspace, req ctrl.Request) (result ctrl.Result, err error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "handleWorkspaceInit")
+	defer tracing.FinishSpan(span, &err)
+
+	if c := wsk8s.GetCondition(ws.Status.Conditions, string(workspacev1.WorkspaceConditionContentReady)); c == nil {
+		var init csapi.WorkspaceInitializer
+		err = proto.Unmarshal(ws.Spec.Initializer, &init)
+		if err != nil {
+			err = fmt.Errorf("cannot unmarshal initializer config: %w", err)
+			return ctrl.Result{}, err
+		}
+
+		alreadyInit, failure, err := wsc.operations.InitWorkspaceContent(ctx, InitContentOptions{
+			Meta: WorkspaceMeta{
+				Owner:       ws.Spec.Ownership.Owner,
+				WorkspaceId: ws.Spec.Ownership.WorkspaceID,
+				InstanceId:  ws.Name,
+			},
+			Initializer: &init,
+			Headless:    ws.Status.Headless,
+		})
+
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("could not initialize workspace %s: %w", ws.Name, err)
+		}
+
+		if alreadyInit {
+			return ctrl.Result{}, nil
+		}
+
+		err = retry.RetryOnConflict(retryParams, func() error {
+			var workspace workspacev1.Workspace
+			if err := wsc.Get(ctx, req.NamespacedName, &workspace); err != nil {
+				return err
+			}
```

Comment:
If the workspace CR has been modified in the meantime by someone else the update will be rejected (the CR generation will be lower than what is saved in etcd) . That's why we need to fetch the workspace again before we update (and after each failed attempt). It is a good point though, I will add a comment to make this clearer.

Reply:
yeah - what i mean is that instead of storing the newer version in the locally scoped `workspace` var, we should load the new version in `ws`, such that when the outer function returns, anything that follows is also immediately using the newer version of the resource

Reply:
D'oh! Of course. I will change it. :smile:

Reply:
👍